src/extractor/worker.ts

```diff
@@ -0,0 +1,98 @@
+import type { Deferred } from '../utils/deferred';
+
+import { extname } from 'path';
+import { Worker, isMainThread, parentPort, workerData } from 'worker_threads';
+import { readFile } from 'fs/promises';
+
+import { loadModule } from '../utils/moduleLoader';
+import { createDeferred } from '../utils/deferred';
+
+export type FileParams = { extractor: string; file: string };
+export type CodeParams = { extractor: string; code: string };
+export type WorkerParams = FileParams | CodeParams;
+
+const IS_TS_NODE = extname(__filename) === '.ts';
+
+// --- Worker functions
+
+let loadedExtractor: string;
+let extractor: Function;
+
+async function loadExtractor(extractorScript: string) {
+  const mdl = await loadModule(extractorScript);
+  extractor = mdl.default;
+}
+
+async function handleJob(args: WorkerParams) {
+  if (loadedExtractor !== args.extractor) {
+    await loadExtractor(args.extractor);
+  }
+
+  const code = 'file' in args ? await readFile(args.file, 'utf8') : args.code;
+
+  return extractor(code);
+}
+
+async function workerInit() {
+  parentPort!.on('message', (msg) => {
+    handleJob(msg.data)
+      .then((res) => parentPort!.postMessage({ id: msg.id, data: res }))
+      .catch((e) => parentPort!.postMessage({ id: msg.id, err: e }));
+  });
+}
+
+if (!isMainThread) workerInit();
+
+// --- Main thread functions
+
+let queryId = 0;
+let worker: Worker | null;
+const queries: Map<number, Deferred<string[]>> = new Map();
+
+function clearWorker() {
+  worker = null;
+  for (const deferred of queries.values()) {
+    deferred.reject('aborted');
+  }
+  queries.clear();
+}
+
+function getWorker(): Worker {
+  if (!worker) {
```

Comment:
Would be cool to run it as many times as the computer has CPU cores to really make it parallell.

Reply:
I'm unsure if concurrency is truly required, as it'd make it harder to keep track of all the states internally for a probably ridiculous gain of a dozen milliseconds. Might need experimentation to confirm 🧪